server/src/com/thoughtworks/go/server/websocket/JobInstanceStatusMonitor.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.websocket;
+
+import com.thoughtworks.go.domain.JobInstance;
+import com.thoughtworks.go.domain.JobResult;
+import com.thoughtworks.go.server.domain.JobStatusListener;
+import com.thoughtworks.go.server.service.JobInstanceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JobInstanceStatusMonitor {
+
+    @Autowired
+    public JobInstanceStatusMonitor(final AgentRemoteHandler agentRemoteHandler, JobInstanceService jobInstanceService) {
+        jobInstanceService.registerJobStateChangeListener(new JobStatusListener() {
+            @Override
+            public void jobStatusChanged(JobInstance job) {
+                if (job.isRescheduled() || JobResult.Cancelled.equals(job.getResult())) {
+                    agentRemoteHandler.sendCancelMessage(job.getAgentUuid());
```

Comment:
Well, that TODO comment is a bit mis-leading. JobResultListener comes into picture for stage cancellations too, provided the job getting cancelled as a part of the stage has an agent assigned to it. Moreover, the class name JobResultListener itself is mis-leading, but thats the right listener which gets invoked when a job/stage gets cancelled either by a user through UI/API or a job-timeout happens. 

Thanks for pointing it out, I did not pay much attention to the reschedule check earlier. A job reschedule happens when the agent responsible for building the job has gone off the radar. In such a case, `AgentRemoteHandler` would probably not find any matching agents in agentSessions, and so a notification would not be sent anyway.

>> I want to make sure we have lowest risk to switch to (or try out) websocket communication in production.

Fair enough point. May be we should register (ie. `jobInstanceService.registerJobStateChangeListener`) only if websocket communication is turned on.